File: ecol/graph.go
package main

type Graph struct {
	n         int
	edge_data [][]int
}

type Edge struct {
	i int
	j int
}

func NewGraph(n int) *Graph {
	edge_data := make([][]int, n)
	for i := 0; i < n; i++ {
		edge_data[i] = make([]int, n)
		for j := 0; j < n; j++ {
			edge_data[i][j] = -1
		}
	}
	return &Graph{n, edge_data}
}

func (g *Graph) Get(i, j int) int {
	return g.edge_data[i][j]
}

func (g *Graph) Set(i, j, v int) {
	g.edge_data[i][j] = v
	g.edge_data[j][i] = v
}

func (g *Graph) EdgeCount() int {
	n := 0
	for i := 0; i < g.n; i++ {
		for j := i; j < g.n; j++ {
			if g.edge_data[i][j] != -1 {
				n++
			}
		}
	}
	return n
}

func (g *Graph) Edges() []Edge {
	edges := []Edge{}
	for i := 0; i < g.n; i++ {
		for j := i; j < g.n; j++ {
			if g.edge_data[i][j] != -1 {
				edges = append(edges, Edge{i, j})
			}
		}
	}
	return edges
}

func (g *Graph) Degree(u int) int {
	d := 0
	for i := 0; i < g.n; i++ {
		if g.edge_data[u][i] != -1 {
			d++
		}
	}
	return d
}

func (g *Graph) CopyInto(h *Graph) {
	for i, x := range g.edge_data {
		for j, v := range x {
			h.edge_data[i][j] = v
		}
	}
}

func (g *Graph) ResetColours() {
	for i := 0; i < g.n; i++ {
		for j := 0; j < g.n; j++ {
			if g.edge_data[i][j] != -1 {
				g.edge_data[i][j] = 0
			}
		}
	}
}

File: pyecol/graph.py
import random


# optimised undirected graph class
class Graph:
    __slots__ = ('n', 'edge_data')

    def __init__(self, n):
        self.n = n
        self.edge_data = [[False] * n for _ in range(n)]

    def __getitem__(self, edge):
        x, y = edge
        data = self.edge_data[x][y]
        if data is False:
            raise KeyError
        return data

    def __setitem__(self, edge, data):
        x, y = edge
        self.edge_data[x][y] = data
        self.edge_data[y][x] = data

    def __delitem__(self, edge):
        x, y = edge
        self[x, y] = False

    def copy(self):
        g = Graph(self.n)
        g.edge_data = [row[:] for row in self.edge_data]
        return g

    def nodes(self):
        return range(self.n)

    def degree(self, node):
        return sum(1 for x in self.edge_data[node] if x is not False)

    def num_edges(self):
        n = 0
        for i in range(self.n):
            row = self.edge_data[i]
            for j in range(i+1, self.n):
                if row[j] is not False:
                    n += 1
        return n

    def edges(self):
        for i in range(self.n):
            row = self.edge_data[i]
            for j in range(i+1, self.n):
                if row[j] is not False:
                    yield (i, j)

    def neighbours(self, node):
        for i, data in enumerate(self.edge_data[node]):
            if data is not False:
                yield i

    def subgraph(self, nodes):
        nl = list(nodes)
        ns = set(nl)
        g = Graph(len(ns))
        for u, v in self.edges():
            if u in ns and v in ns:
                i = nl.index(u)
                j = nl.index(v)
                g[i, j] = self[u, v]
        return g

    def degrees(self):
        return {u: self.degree(u) for u in self.nodes()}


def complete_graph(k):
    g = Graph(k)
    for i in range(k):
        for j in range(i+1, k):
            g[i, j] = 0
    return g


def erdos_renyi_graph(n, p):
    r = random.Random()
    g = Graph(n)
    for i in range(n):
        for j in range(i+1, n):
            if r.random() <= p:
                g[i, j] = 0
    return g


def complete_bipartite_graph(n, m):
    G = Graph(n + m)
    for i in range(n):
        for j in range(m):
            G[i, n + j] = 0
    return G
